[c10d] Add support for testing SIGABRT return

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #153167

`SIGABRT` is a common return by *negative* distributed tests, which checks for effectiveness of NaN assert, watchdog throw, etc.

These errors are not detectable by traditional statements like `with self.assertRaises(RuntimeError)`.

Instead, we'd need to check for the process's return code, e.g. `SIGABRT(6)` would have a return code of -6.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

Differential Revision: [D75467002](https://our.internmc.facebook.com/intern/diff/D75467002)

kwen2501 committed:
[c10d] Add support for testing SIGABRT return

[ghstack-poisoned]

kwen2501 commented on torch/testing/_internal/common_distributed.py:
```diff
@@ -838,28 +846,13 @@ def _join_processes(self, fn) -> None:
             for pipe in self.pid_to_pipe.values():
                 pipe.close()
 
-    def _check_no_test_errors(self, elapsed_time) -> None:
-        """
-        Checks that we didn't have any errors thrown in the child processes.
-        """
-        for i, p in enumerate(self.processes):
-            if p.exitcode is None:
-                raise RuntimeError(
-                    f"Process {i} timed out after {elapsed_time} seconds"
-                )
-            self.assertNotEqual(self.TEST_ERROR_EXIT_CODE, p.exitcode)
```
This function is merged into `_check_return_codes`

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -573,16 +573,13 @@ def test_nan_rank_filter(self):
-        x = torch.ones((10,), dtype=torch.bfloat16, device=device) * self.rank
-        t = torch.ones(3, 4, dtype=torch.bfloat16, device=device)
+        x = torch.ones((10,), device=device) * self.rank
+        t = torch.ones(3, 4, device=device)
```
so that we run the test without SM limitation (bfloat16 requires 80 and up)

kwen2501 committed:
Update on "[c10d] Add support for testing SIGABRT return"


`SIGABRT` is a common return by *negative* distributed tests, which checks for effectiveness of NaN assert, watchdog throw, etc.

These errors are not detectable by traditional statements like `with self.assertRaises(RuntimeError)`.

Instead, we'd need to check for the process's return code, e.g. `SIGABRT(6)` would have a return code of -6.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

fduwjj reviewed (approved):
stamp to unblock

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -531,14 +531,14 @@ def test_nan_assert(self, type):
+        pg._allgather_base(output, nan_tensor)
 
         backend._set_enable_nan_check(True)
-        with self.assertRaises(RuntimeError):
```
Yes

kwen2501 commented:
@pytorchbot merge

cyyever commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot revert -m "Broke lint, see https://hud.pytorch.org/hud/pytorch/pytorch/fe784c5a2cea1a811d2489a2d3c635c14e4253fe/1?per_page=50&name_filter=lintrunner-noclang&mergeEphemeralLF=true" -c nosignal

kwen2501 committed:
Update on "[c10d] Add support for testing SIGABRT return"


`SIGABRT` is a common return by *negative* distributed tests, which checks for effectiveness of NaN assert, watchdog throw, etc.

These errors are not detectable by traditional statements like `with self.assertRaises(RuntimeError)`.

Instead, we'd need to check for the process's return code, e.g. `SIGABRT(6)` would have a return code of -6.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge -i

malfet commented:
@pytorchbot revert -m "It broke lint" -c nosignal

nWEIdia commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -44,13 +44,11 @@
     requires_nccl_version,
     skip_if_lt_x_gpu,
     skip_if_rocm_multiprocess,
-    sm_is_or_higher_than,
```
Please rebase with the latest main and you can see this cannot be removed due to it still being used.

kwen2501 replied:
Oh no wonder it kept breaking trunk but my local linter did not show any error!
Thank you for the heads-up! @nWEIdia 
Sorry about the inconvenience! @malfet 
I am rebasing against main now.

nWEIdia reviewed (dismissed):
A recent change was still invoking sm_is_or_higher_than. So this cannot be removed.

kwen2501 committed:
Update on "[c10d] Add support for testing SIGABRT return"


`SIGABRT` is a common return by *negative* distributed tests, which checks for effectiveness of NaN assert, watchdog throw, etc.

These errors are not detectable by traditional statements like `with self.assertRaises(RuntimeError)`.

Instead, we'd need to check for the process's return code, e.g. `SIGABRT(6)` would have a return code of -6.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 commented:
@pytorchbot merge

nWEIdia reviewed (commented):
test_nan_assert was skipped via the following: 

https://github.com/pytorch/pytorch/pull/153167/files#diff-9f639571a250cffbe9cded7d2fbb5ad6311e4be9c0c7610e5ba85930806e7f38R485 

It would be great if we test whether the change now works better than pg.allreduce. 
But we need to unskip this first, because I used  not ( .... and false) which would always skip the test.

kwen2501 commented:
@kwen2501 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75467002).
